### PR TITLE
feat(agent): publish direct Convex REST API

### PR DESCRIPTION
**Pull request by nabilfatih**
## Summary

Adds Nakafa's protected read-only REST v1 successor directly to Convex HTTP Actions at `/internal/agent`, with OpenAPI 3.1.1, typed failures, quotas, and strict signed-content reads. The deployed predecessor at `/v1` remains unchanged, so public Vercel traffic is not switched by this PR.

## Why

Convex owns the signed publication and read models. A direct Convex HTTP boundary removes the extra Next.js execution hop after a separately verified Vercel bridge switch, while Vercel retains the public domain, firewall, trusted client-address boundary, and rotating-secret injection.

## Scope

- Adds index, health, search, content, taxonomy, Quran, and OpenAPI endpoints.
- Derives runtime and OpenAPI contracts from Effect schemas.
- Requires the Vercel-owned origin secret and rejects direct origin access.
- Uses the official Convex rate limiter under an explicit Nakafa quota policy.
- Reads focused Quran content and bounded article taxonomy through single internal query transactions.
- Preserves citation-only material topics without advertising nonexistent markdown.
- Logs private unexpected causes with the same sanitized public request ID.

## Exact-head verification

Exact head `ff73a83215005e86d6dac1df818f95692ea61e76` is based on protected main `4e1c1ff3d62ec29a0a5aa3bbd63d9d70e3635194`. Range-diff preserves all 11 PR patches; the only context change is main's independently merged `@effect/vitest` dependency.

Exact-head CI run 33000548841 passed Quality, signed Production acceptance, Browser, agent documentation, runtime verification, cleanup, and Required. Doctor run 33000548900 passed. Frozen install, 30 focused owner and route tests, backend typecheck, diff check, and worktree cleanliness are green. The patch-equivalent full proof also passed backend 335 files and 1,432 tests, contents 21 files and 139 tests at 100 percent coverage, both workspace typechecks, lint, boundaries, security audit, Effect source identity, isolated Convex codegen, and EN, ID, DE taxonomy bounds. Fresh review comment 5429888573 reviewed `ff73a83215` and found no major issue; all seven threads are resolved.

## Dependency and rollout state

This additive PR does not switch traffic, create a Preview, manually deploy, or write production data. After protected merge, exact-main Convex and Vercel production evidence must prove the protected origin before separate MCP expansion and Vercel bridge switch PRs.

## Material risks

The exact head is unchanged, clean, mergeable, and fully gated. The later switch must inject the secret, preserve Vercel's documented overwritten `x-forwarded-for`, disable external rewrite caching, and pass live acceptance. Temporary Next transports remain only through migration and must be deleted with their adapters, dependencies, and tests after the switch gate.

### CODE REVIEW

**Comment by nabilfatih** [comment]
@codex review

Please review exact head `8be2b9cea4e3c5e6eb5949000295316db07f8676`. Focus on security, direct Convex HTTP Action behavior, Effect error boundaries, signed-content release consistency, OpenAPI/runtime parity, rate-limit semantics, and deletion-ready rollout architecture. Treat existing Next transports as temporary until the separate verified switch and contraction.

File: packages/backend/agent/edge.ts [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+export type ApiEdgeContract = typeof ApiEdgeContractSchema.Type;
+
+/** Server-only contract shared by the Vercel bridge and Convex origin. */
+export const NAKAFA_API_EDGE_CONTRACT: ApiEdgeContract = {
+  secretEnvironment: API_EDGE_SECRET_ENVIRONMENT,
+  secretHeader: API_EDGE_SECRET_HEADER,
+};
+
+/** Trusted Vercel-provided client address used for pseudonymous quotas. */
+export const NAKAFA_EDGE_CLIENT_IP_HEADER = "x-forwarded-for";
```
[the remark this answers is not in the document]
Comment (nabilfatih) [reply packages/backend/agent/edge.ts:23]: Independently rejected with primary evidence. Vercel documents that it overwrites `x-forwarded-for` and does not forward external IPs specifically to prevent spoofing, except when an Enterprise trusted proxy is enabled: https://vercel.com/docs/headers/request-headers. Nakafa's Vercel team is on Pro, and the planned bridge has no trusted-proxy mode. `x-vercel-forwarded-for` is not a documented Vercel request header. The origin also fails closed when `x-forwarded-for` is absent. Keeping the documented overwritten header is therefore the source-backed behavior; the bridge acceptance test will verify that exact forwarded value before public switch.

File: packages/backend/convex/routes/agent/api.ts [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+  program: Effect.Effect<Response, AgentDomainError>
+) {
+  const instance = new URL(request.url).pathname;
+  return Effect.runPromise(
+    program.pipe(
+      Effect.matchCause({
+        onFailure: (cause) => {
+          const failure = cause.reasons.find(Cause.isFailReason);
+          if (!failure) {
+            return internalFailureResponse(instance, requestId);
```
[the remark this answers is not in the document]
Comment (nabilfatih) [reply packages/backend/convex/routes/agent/api.ts]: Accepted. Commit `77ac70088` changes the HTTP boundary to `Effect.matchCauseEffect` and logs the complete unexpected `Cause` with the same `requestId` and `instance` before returning the sanitized 500. `response.test.ts` proves the private defect is present in the structured error log, the correlation fields match the response, and the private detail is absent from the RFC 9457 body. The exact agent API suite passes 21/21.

File: packages/backend/agent/content.ts [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+  if (Option.isNone(ref)) {
+    return yield* contentReadError(
+      "The signed content reference has an invalid graph identity."
+    );
+  }
+  if (ref.value.section === "quran") {
+    return yield* readQuranMarkdown(ctx, ref.value);
+  }
+  if (isPublishedRef(ref.value)) {
+    return yield* readPublishedMarkdown(ctx, ref.value);
```
[the remark this answers is not in the document]
Comment (nabilfatih) [reply packages/backend/agent/content.ts:95]: Accepted after tracing the signed material model. A topic is a citation/grouping identity and owns no `subject-lesson` runtime artifact, while the shared builder inferred markdown capability from `section: material`. Commit `e3b67eae9` replaces that inference with required `hasMarkdownSource`, emits no `markdown_url` for topics and try-outs, preserves the capability through decoded summaries, and requires it before article/material runtime dispatch. Focused proof is green: 7 backend files and 46 tests, including direct topic reference and `/v1/content` behavior, plus the contents reference suite and both workspace typechecks.

File: packages/backend/convex/routes/agent/api.ts [hunk 4]
```diff
@@ -0,0 +1,289 @@
...
+  | AgentRateLimitError
+  | NakafaAgentDataReadError
+  | NakafaAgentInputError;
+type AgentApp = HonoWithConvex<ActionCtx, { requestId: string }>;
+
+/** Registers the protected read-only API and its machine-readable contract. */
+export function registerAgentApiRoutes(app: AgentApp) {
+  app.use("/openapi.json", guardAgentApi);
+  app.use("/v1", guardAgentApi);
+  app.use("/v1/*", guardAgentApi);
```
[the remark this answers is not in the document]
Comment (nabilfatih) [reply packages/backend/convex/routes/agent/api.ts]: Accepted after tracing the installed `@convex-dev/better-auth` 0.12.5 handler: the current Next route forwards the original `/v1` path to Convex and does not inject `x-nakafa-api-edge-secret`. Commit `d2cc05bf0` therefore makes this a real expand phase. It keeps the deployed `/v1` predecessor unchanged and mounts the protected successor at `/internal/agent`, with public problem instances projected back to `/v1`. Isolated Agent Mode proof shows predecessor 200, direct successor 403, authenticated successor 200, OpenAPI 3.1.1, and invalid ref 422. The predecessor remains owned by the later verified Vercel switch and contraction.

File: packages/backend/agent/content.ts [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+  }
+  const reference = yield* readAgentQuery(
+    ctx,
+    contentReference,
+    { input: lookup.value },
+    "Unable to resolve the Nakafa content reference."
+  );
+  if (!reference) {
+    return Option.none<NakafaAgentMarkdown>();
+  }
```
[the remark this answers is not in the document]
Comment (nabilfatih) [reply packages/backend/agent/content.ts:74]: Accepted. Convex documents that separate `ctx.runQuery` calls are separate transactions and can observe different state. Commit `8af695ad0` adds one internal query whose domain implementation resolves the reference and reads Quran markdown inside the same Convex query transaction; `879a320df` isolates the registered adapter in folder-owned `reference/internal.ts`. Focused unit and HTTP proof covers the combined Quran source, and backend typecheck is green. Source: https://docs.convex.dev/functions/actions#avoid-await-ctx-runmutation--await-ctx-runquery

File: packages/backend/convex/routes/agent/api.ts [hunk 6]
```diff
@@ -0,0 +1,289 @@
...
+    )
+  );
+
+  app.get("/v1/content", (context) =>
+    runMeteredRequest(
+      context.env,
+      context.req.raw,
+      context.get("requestId"),
+      readContentInput(new URL(context.req.url)).pipe(
+        Effect.flatMap((ref) => getNakafaContent(context.env, ref)),
```
[the remark this answers is not in the document]
Comment (nabilfatih) [reply packages/backend/convex/routes/agent/api.ts:142]: Accepted. `readContentInput` previously enforced only presence, while the OpenAPI-owned `NakafaAgentContentRefInputSchema` requires a canonical Nakafa URL, graph content ID, or resource URI. Commit `d2cc05bf0` decodes that exact schema before lookup. Focused API proof now returns 422 for both an empty ref and a non-Nakafa URL, while valid missing references remain 404.

File: packages/backend/agent/taxonomy.ts [hunk 7]
```diff
@@ -0,0 +1,290 @@
...
+    ) {
+      return yield* missingInventory("article taxonomy", locale);
+    }
+    if (page.result.isDone) {
+      return categories;
+    }
+    if (page.result.continueCursor === "") {
+      return yield* missingInventory("article taxonomy cursor", locale);
+    }
+    return yield* readCategoryPages(ctx, locale, {
```
[the remark this answers is not in the document]
Comment (nabilfatih) [reply packages/backend/agent/taxonomy.ts]: Valid. I reproduced that the action recursively called `contentRelease/article:categories`, so each 32-row page was a separate Convex transaction and the accumulated array had no total bound. Convex explicitly recommends batching repeated `runQuery` reads into one internal query because separate calls are separate transactions: https://docs.convex.dev/functions/actions#avoid-await-ctxrunmutation--await-ctxrunquery. Convex also caps one query transaction at 16 MiB read: https://docs.convex.dev/production/state/limits#transactions.

Commit e62e40940f replaces the recursion with `contentRelease/article/internal:readAgentTaxonomy`, one indexed and release-verified query. Its 128-category ceiling is derived from the repository's existing transaction proof, `floor((16 MiB - 4 MiB headroom) / (6 reads per route * 16 KiB per read model))`, not a guessed API limit. It fails closed before verification on row 129 and never returns partial taxonomy. Focused proof: EN, ID, and DE complete reads, unmanaged state, success at exactly 128, failure at 129, 30 agent owner/route tests green, backend typecheck green, and isolated Convex function-spec exposes exactly one internal query registration.

**Comment by nabilfatih** [comment]
@codex review
